### ISSUE TITLE
Tooltip initialized with on="click" can only be closed by click on another data point

### ISSUE THREAD

**Issue by paulklemm** [rstudio/ggvis#250]
This issue is related to this stackoverflow question: [https://stackoverflow.com/questions/25619778/hide-tooltip-initialized-with-on-click-in-ggvis-plot](https://stackoverflow.com/questions/25619778/hide-tooltip-initialized-with-on-click-in-ggvis-plot)

The behavior of tooltips initialized with on="click" is not intuitive in my opinion. A tooltip can only be closed by clicking on another data point in the plot. I would expect the following behavior:
- close the tooltip on click of the data point, which triggered it
- close the tooltip on click of another data point and open the respective tooltip (the current functionality)
- close the tooltip on click on anywhere in the plot outside of the tooltip (still makes it possible to interact with content of the tooltip, e.g. copy text)

**Comment by jhtravis** [comment]
Looking forward to this feature as well.  I am planning to use links in my tooltip which will require a delayed (or customizable) hover or a clickable link that can exited by clicking on a black bar of the graph.  Awesome work on the package.

**Comment by daattali** [comment]
Until this gets impemented, if you want to have this behaviour in a shiny app you can easily achieve it with minimal javascript. Example:

```
library(shiny)
library(ggvis)

jscode <- 
"$(function() {
  $('#ggvis').click(function(){ $('#ggvis-tooltip').hide(); });
})
"

shinyApp(
  ui = fluidPage(
    tags$script(jscode),
    uiOutput("ggvis_ui"),
    ggvisOutput("ggvis")
  ),
  server = function(input, output, session) {
    mtcars %>% 
      ggvis(~wt, ~mpg) %>%
      layer_points() %>%
      add_tooltip(function(df) df$wt, on = "click") %>%
      bind_shiny("ggvis", "ggvis_ui")
  }
)
```

**Comment by paulklemm** [comment]
@daattali Really nice hack, works well for shiny apps. I use `ggvis` as [shiny reactive](https://rstudio.github.io/shiny/tutorial/#reactivity) in an [interactive RMarkdown document](http://rmarkdown.rstudio.com/authoring_shiny.html). I guess I can [convert the reactive into a shiny app](http://rmarkdown.rstudio.com/authoring_embedded_shiny.html). Is there, however, a way to inject the javascript into shiny reactives?

**Comment by daattali** [comment]
Sorry I'm not quite sure what you mean by injecting JS into a reactive

**Comment by oganm** [comment]
I think you are trying to add the javascript command to the markdown document. If that is the case just adding <script>$(function () { $("#ggvis").click(function(e){ $("#ggvis-tooltip").hide(); }); })</script> to somewhere in the document should to it

**Comment by paulklemm** [comment]
@oganm You are right. I've added the code to the document, but it does not work. @oganm @daattali Any idea whats wrong here? Here is a (rather stupid) minimal example:

    ---
    title: "ggvis Hide Tooltip on Click"
    runtime: shiny
    output: 
      html_document
    ---
    
    <script>
    $(function() {
      $('#ggvis').click(function(){ $('#ggvis-tooltip').hide(); });
    })
    </script>
    
    ```{r loading_main_script, echo=FALSE, results='hide',message=FALSE, warning=FALSE, cache=FALSE}
    library(ggvis)
    # Define Shiny reactive
    plot_data <- reactive({
      
      mtcars %>% ggvis(~wt, ~mpg) %>% 
      layer_points() %>% 
      add_tooltip(function(df) df$wt, on = "click")
    })
    
    bind_shiny(plot_data, 'plot_data')
    ```
    
    ```{r echo=FALSE,message=FALSE, warning=FALSE, error=FALSE, out.extra=FALSE}
    
    ggvisOutput('plot_data')
    
    ```

The `<script>` tag should work according to [this stackoverflow post](https://stackoverflow.com/questions/25819539/how-to-add-an-interactive-visualization-to-r-markdown/25921889#25921889).

**Comment by daattali** [comment]
It's not working because you changed the name of the element. My code works by running the JS code whenever you click on the element with id `ggvis`, but in your code there is no such element, it's called `plot_data` instead. So change the JS function to ` $('#plot_data')` instead of ` $('#ggvis')`

**Comment by paulklemm** [comment]
Thanks a ton, works like a charm! If you want to, you can add this as answer to [my stackoverflow question](https://stackoverflow.com/questions/25619778/hide-tooltip-initialized-with-on-click-in-ggvis-plot) for fame and glory :wink: .

**Comment by daattali** [comment]
:+1: responses, thanks

**Comment by speedreader** [comment]
Strange, using this verbatim I can't seem to get it to work. The id I'm passing to 'ggvisOutput()' is 'ggvis'. But I'm getting the following: 

'Warning: Error in tag: object 'jscode' not found'

Does it matter that I'm trying to use the tooltip from reactive output?

**Comment by daattali** [comment]
@speedreader based on that error, it looks like you aren't defining the `jscode` variable, or you're defining it with an error
